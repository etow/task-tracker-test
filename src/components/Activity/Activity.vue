<template>
  <div class="activity">
    <div v-for="[name] in activity" :key="name" class="activity__item">
      <div class="activity__key">{{name}}</div>
      <div class="activity__time">{{ getTimeElapsed(name) }}</div>
    </div>
    <div v-if="isDone" class="activity__item is-done">
      {{ taskToEdit.category }} âœ“
    </div>
  </div>
</template>
<script>
export default {
  computed: {
    isDone() {
      return this.categories[this.taskToEdit.category]?.endOfWorkflow;
    },
    taskToEdit() {
      return this.$store.state.taskToEdit;
    },
    activityData() {
      return this.taskToEdit.activity || {};
    },
    activity() {
      return Object.entries(this.activityData)
        .filter(([key]) => this.categories[key].endOfWorkflow !== true);
    },
    categories() {
      return this.$store.state.categories;
    },
  },
  methods: {
    getTimeElapsed(categoryName) {
      // collect records
      const time = this.activityData[categoryName].reduce((acc, entry) => {
        const started = entry.started;
        const finished = entry.finished || Date.now();
        acc += (finished - started);
        return acc;
      }, 0);

      // conver time to hours and minutes
      const minutes = Math.floor((time / (1000 * 60)) % 60);
      const hours = Math.floor((time / (1000 * 60 * 60)) % 24);

      let output = '';
      if (hours > 0) output += `${hours} ${(hours > 1 || hours === 0) ? 'hours' : 'hour'}, `;
      output += `${minutes} ${(minutes > 1 || minutes === 0) ? 'minutes' : 'minute'}`;

      return output;
    },
  }
};
</script>
<style lang="scss" scoped>
.activity {
  &__item {
    display: flex;
    height: 30px;
    line-height: 30px;
    border-bottom: 1px solid rgb(228, 231, 237);
    &.is-done {
      text-align: center;
      display: block;
      border-bottom: 0;
    }
  }

  &__key {
    min-width: 100px;
  }
  &__time {
    padding: 0 20px;
  }
}
</style>
