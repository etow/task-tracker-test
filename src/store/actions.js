import taskRepository from '../repositories/taskRepository';
import taskCategoryRepository from '../repositories/taskCategoryRepository';
import {
  SET_TASKS_CATEGORIES,
  SET_TASKS_CATEGORIES_DATA,
  SET_TASKS,
  ADD_TASK,
  DELETE_TASK,
  SET_TASK_TO_EDIT,
  SET_TASK,
  FETCH_TASKS_CATEGORIES,
  FETCH_TASKS,
  CREATE_TASK,
  UPDATE_TASK,
  ROLLBACK_STATE,
  BACKUP_STATE,
  UPDATE_TASKS,
} from './constants';

export default {
  [FETCH_TASKS_CATEGORIES]: ({ commit }) => {
    return taskCategoryRepository.get()
      .then((response) => {
        commit(SET_TASKS_CATEGORIES, response.data);
        commit(SET_TASKS_CATEGORIES_DATA, response.data);
      });
  },
  [FETCH_TASKS]: ({ commit }) => {
    return taskRepository.get()
      .then((response) => commit(SET_TASKS, response.data));
  },
  [CREATE_TASK]: ({ commit, state }, newTask) => {
    commit(BACKUP_STATE, state);
    const task = {
      ...newTask,
      id: Math.random().toString(36).substr(2, 6),
      order: state.tasks[newTask.category].length,
    };
    commit(ADD_TASK, task);
    return taskRepository.create(task);
  },
  [DELETE_TASK]: ({ commit, state }, {task, category}) => {
    commit(BACKUP_STATE, state);
    commit(DELETE_TASK, {task, category});
    commit(SET_TASK_TO_EDIT, {});
    return taskRepository.delete(task.id);
  },
  [UPDATE_TASK]: ({ commit, state }, {task, prevCategory}) => {
    commit(BACKUP_STATE, state);
    if (prevCategory && task.category !== prevCategory) {
      commit(DELETE_TASK, { task, category: prevCategory });
    }
    const targetIndex = state.tasks[task.category].findIndex((aTask) => aTask.id === task.id);
    if (targetIndex >= 0) {
      commit(SET_TASK, { task, targetIndex });
    } else {
      const order = state.tasks[task.category].length;
      commit(ADD_TASK, { ...task, order });
    }
    commit(SET_TASK_TO_EDIT, {})
    return taskRepository.update(task);
  },
  [UPDATE_TASKS]: ({ commit, state }, { tasks, category }) => {
    commit(BACKUP_STATE, state);
    tasks.forEach((task, index) => {
      task.category = category;
      task.order = index;
    });
    commit(UPDATE_TASKS, { tasks, category });

    return taskRepository.update(tasks);
  },
  [SET_TASK_TO_EDIT]: ({ commit, state}, task) => {
    commit(BACKUP_STATE, state);
    commit(SET_TASK_TO_EDIT, task);
  },
  [ROLLBACK_STATE]: ({ commit }) => {
    commit(ROLLBACK_STATE);
  },
};
