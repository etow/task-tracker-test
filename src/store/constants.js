

/* mutations and actions */
export const ADD_TASK = 'ADD_TASK';
export const SET_TASKS = 'SET_TASKS';
export const SET_TASKS_CATEGORIES = 'SET_TASKS_CATEGORIES';
export const SET_TASKS_CATEGORIES_DATA = 'SET_TASKS_CATEGORIES_DATA';
export const SET_TASK = 'SET_TASK';
export const SET_TASK_TO_EDIT = 'SET_TASK_TO_EDIT';
export const DELETE_TASK = 'DELETE_TASK';
export const FETCH_TASKS_CATEGORIES = 'FETCH_TASKS_CATEGORIES';
export const FETCH_TASKS = 'FETCH_TASKS';
export const CREATE_TASK = 'CREATE_TASK';
export const UPDATE_TASK = 'UPDATE_TASK';
export const ROLLBACK_STATE = 'ROLLBACK_STATE';
export const BACKUP_STATE = 'BACKUP_STATE';
export const UPDATE_TASKS = 'UPDATE_TASKS';
export const UPDATE_TASK_ACTIVITY = 'UPDATE_TASK_ACTIVITY';
