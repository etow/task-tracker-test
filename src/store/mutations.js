import {
  ADD_TASK,
  SET_TASKS,
  SET_TASKS_CATEGORIES,
  SET_TASKS_CATEGORIES_DATA,
  SET_TASK,
  SET_TASK_TO_EDIT,
  DELETE_TASK,
  ROLLBACK_STATE,
  BACKUP_STATE,
  UPDATE_TASKS,
  UPDATE_TASK_ACTIVITY
} from './constants';

export default {
  [ADD_TASK]: (state, task) => {
    state.tasks[task.category].push(task);
  },
  [SET_TASKS]: (state, tasks) => {
    Object.keys(state.tasks)
      .forEach((category) => {
        const tasksByCategory = tasks.filter((task) => task.category === category);
        state.tasks[category] = tasksByCategory.sort((a, b) => a.order - b.order);
      });
  },
  [SET_TASKS_CATEGORIES]: (state, categories) => {
    categories.forEach((cat) => state.tasks[cat.name] = []);
  },
  [SET_TASKS_CATEGORIES_DATA]: (state, categories) => {
    categories.forEach((cat) => state.categories[cat.name] = cat);
  },
  [SET_TASK]: (state, { task, targetIndex }) => {
    const index = targetIndex || state.tasks[task.category].findIndex((aTask) => aTask.id === task.id);
    state.tasks[task.category][index] = task;
  },
  [SET_TASK_TO_EDIT]: (state, task) => {
    state.taskToEdit = task;
  },
  [DELETE_TASK]: (state, {task, category}) => {
    state.tasks[category] = state.tasks[category].filter((aTask) => aTask.id !== task.id);
  },
  [ROLLBACK_STATE]: (state) => {
    Object.assign(state, JSON.parse(JSON.stringify(state.prevState)));
  },
  [BACKUP_STATE]: (state, currentState) => {
    state.prevState = JSON.parse(JSON.stringify(currentState));
  },
  [UPDATE_TASKS]: (state, {tasks, category}) => {
    state.tasks[category] = tasks;
  },
  [UPDATE_TASK_ACTIVITY]: (state, tasks) => {
    if (!Array.isArray(tasks)) {
      tasks = [tasks];
    }
    tasks.forEach((task) => {
      const found = state.tasks[task.category].find((aTask) => aTask.id === task.id);
      found.activity = task.activity;
    });
  }
};
