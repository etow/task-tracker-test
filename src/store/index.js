import { createStore } from 'vuex';

import mutations from './mutations';
import actions from './actions';

export default createStore({
  state: {
    tasks: {},
    categories: {},
    taskToEdit: {},
    prevState: {},
  },
  mutations,
  actions,
});
