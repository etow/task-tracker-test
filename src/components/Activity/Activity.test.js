import { shallowMount } from '@vue/test-utils';
import { createStore } from 'vuex';
import Activity from './Activity.vue';

import { SET_TASK_TO_EDIT } from '../../store/constants';

const tasks = [
  {
    name: 'test1',
    category: 'Planned',
    order: 0,
    estimate: 0,
    activity: {
      Planned: [{ started: 0, finished: 900000 }, { started: 0, finished: 9000000 }],
      'In Progress': [{ started: 0, finished: 600000 }],
      Completed: [{ started: 0, finished: 600000 }],
    },
  },
  {
    name: 'test2',
    category: 'Completed',
    order: 0,
    estimate: 0,
    activity: {
      Planned: [{ started: 0, finished: 900000 }, { started: 0, finished: 9000000 }],
      'In Progress': [{ started: 0, finished: 600000 }],
    },
  },
  {
    name: 'test3',
    category: 'Completed',
    order: 0,
    estimate: 0,
    activity: {
      Planned: [{ started: 0, finished: 900000 }, { started: 0, finished: 9000000 }],
      'In Progress': [{ started: 0 }],
    },
  },
];

describe('Activity.vue', () => {
  it('renders tasks Planned and In Progress', async () => {
    const [store] = mockStore({ taskToEdit: tasks[0] });
    const wrapper = createWrapper((dataToMount) => {
      dataToMount.global.plugins = [store];
      return dataToMount;
    });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('renders tasks Planned and In Progress and Completed', async () => {
    const [store] = mockStore({ taskToEdit: tasks[1] });
    const wrapper = createWrapper((dataToMount) => {
      dataToMount.global.plugins = [store];
      return dataToMount;
    });

    expect(wrapper.html()).toMatchSnapshot();
  });

  describe('getTimeElapsed', () => {
    it('calculates a single activity', async () => {
      const result = Activity.methods.getTimeElapsed.call({
        activityData: {
          Planned: [{
            started: 0, finished: 900000,
          }],
        },
      }, 'Planned');

      expect(result).toEqual('15 minutes');
    });

    it('it calculates the sum of activity', async () => {
      const result = Activity.methods.getTimeElapsed.call({
        activityData: {
          Planned: [
            {
              started: 0, finished: 900000,
            },
            {
              started: 0, finished: 900000,
            },
          ],
        },
      }, 'Planned');

      expect(result).toEqual('30 minutes');
    });

    it('it completes the finished time it it is not finished yet', async () => {

      const dateSpy = jest.spyOn(Date, 'now').mockImplementation(() => 900000);

      const result = Activity.methods.getTimeElapsed.call({
        activityData: {
          Planned: [
            {
              started: 0,
            },
          ],
        },
      }, 'Planned');

      expect(result).toEqual('15 minutes');
      dateSpy.mockRestore();
    });

    it('it calculates hours and minutes', async () => {
      const result = Activity.methods.getTimeElapsed.call({
        activityData: {
          Planned: [
            {
              started: 0, finished: 8100000,
            },
          ],
        },
      }, 'Planned');

      expect(result).toEqual('2 hours, 15 minutes');
    });

    it('it shows hour and minute in singular', async () => {
      const result = Activity.methods.getTimeElapsed.call({
        activityData: {
          Planned: [
            {
              started: 0, finished: 3660000,
            },
          ],
        },
      }, 'Planned');

      expect(result).toEqual('1 hour, 1 minute');
    });
  });
});


const createWrapper = (decorator = (n) => n) => {
  const dataToMount = {
    global: {
      stubs: {
      },
    },
  };

  decorator(dataToMount);
  return shallowMount(Activity, dataToMount);
};

function mockStore({ set_task_to_edit, taskToEdit } = {}) {
  const state = {
    tasks: tasks,
    categories: {
      Planned: { name: 'Planned', color: 'red' },
      'In Progress': { name: 'In Progress', color: 'red' },
      Completed: { name: 'Completed', color: 'red', endOfWorkflow: true },
    },
    taskToEdit: taskToEdit || {},
  };

  const actions = {
    [SET_TASK_TO_EDIT]: set_task_to_edit || jest.fn(() => Promise.resolve()),
  }

  const store = createStore({
    state,
    actions,
  });

  return [store, { actions }];
}
