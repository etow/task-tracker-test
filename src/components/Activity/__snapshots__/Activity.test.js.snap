// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Activity.vue renders tasks Planned and In Progress 1`] = `
<div class="activity">
  <div class="activity__item">
    <div class="activity__key">Planned</div>
    <div class="activity__time">2 hours, 45 minutes</div>
  </div>
  <div class="activity__item">
    <div class="activity__key">In Progress</div>
    <div class="activity__time">10 minutes</div>
  </div>
  <!--v-if-->
</div>
`;

exports[`Activity.vue renders tasks Planned and In Progress and Completed 1`] = `
<div class="activity">
  <div class="activity__item">
    <div class="activity__key">Planned</div>
    <div class="activity__time">2 hours, 45 minutes</div>
  </div>
  <div class="activity__item">
    <div class="activity__key">In Progress</div>
    <div class="activity__time">10 minutes</div>
  </div>
  <div class="activity__item is-done">Completed âœ“ </div>
</div>
`;
